helper/awsutil/region.go

```diff
@@ -0,0 +1,44 @@
+package awsutil
+
+import (
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+)
+
+func GetOrDefaultRegion(configuredRegion string) string {
+	// We default to us-east-1 because it's a widely used region
+	// and is also where AWS first rolls out new features.
+	defaultRegion := "us-east-1"
+
+	// We prefer env variables to configured ones because they
+	// serve as a way to change the application's configuration
+	// on the fly rather than via restarting some process.
+	if os.Getenv("AWS_REGION") != "" {
```

Comment:
I guess it depends on how you define "configured" but I'm not sure this is right.

In general, the way I tend to think of the precedence of AWS picking up configuration is that it's "more specific takes precedence over less specific." For example:
1. Configuration of, say, the region on the command line for a given invocation is the most specific to a request and that takes highest precedence.
2. Environment variables are potentially shared across many invocations and so they have less precedence
3. Configuration in `~/.aws/config` is shared across all invocations of a given user and so this has even less precedence
4. Configuration retrieved from the EC2 instance metadata service is shared by all invocations on a given machine (modulo some extra effort to prevent it from being used), and so it has the lowest precedence.

In the case of the `vault login -method=aws region=<some_region>`, the region is specified on the command line and should take highest precedence according to this model. That's how you would override a given environment variable without unsetting it. But in this implementation, the region passed on the command line is given lower precedence than the environment variable, and I would find that confusing.

On the other hand, I can see how, when you're reading a configuration out of a file (the case of the Vault agent), that's somewhat akin to reading from `~/.aws/config` (case 3 above) and environment variables should take precedence over config specified in a file, and that's built in to the existing awskms seal code, and so changing that would potentially be a breaking change. My gut is that specifying configuration for a non-AWS app (e.g., Vault) in a config file is more akin to explicitly specified configuration (case 1 above), and it should take precedence over environment variables. Like, i'd tend to prefer config specified on the Vault agent's command line, then config specified in a config file for Vault agent (these would be sub-categories under case 1), then fall back to environment variables, then fall back to the AWS config files.

Anyway, I realize this is a lot of text for a relatively small decision, apologies for that.

Reply:
I think that's a totally valid framework! I like it! 

The only reason I did it in this particular order is because it pre-exists this way in `vault/seal/awskms/awskms.go`. Do you think I should update the logic as you describe both here _and_ there, or let the logic diverge?

Reply:
Actually, as I'm coding this up, I'm running into some challenges. The main one is that, as far as I can tell, nothing is exported from the AWS SDK around providing regions. It does have a shared configuration provider, for instance, but it's private in the `session` package. 

I could recreate that code in Vault but then it has the risk of diverging. Or I could use the region returned by `session.New()`, but that won't tell me if it's just defaulting or if it actually found something, which makes it impossible to know whether to also poke around in the instance metadata.

For now, I'm thinking it might make sense to do it like this:
- If configured by user, use the configured value
- If available in env vars, use those
- If available from instance metadata, use that
- Otherwise, default to us-east-1

Reply:
I should elaborate that I was hoping the AWS SDK would export some of that stuff because looking for config files in locations that vary based on OS could get a little tricky.

Reply:
Hmm, yeah, makes sense.

The challenge here, IMO, is trying to provide a consistent experience to people based on what they already experience, but the experiences actually vary in subtle ways across SDKs. For example, boto3, the Python SDK (which the aws cli uses), loads `~/.aws/config` by default and only reads the `AWS_DEFAULT_REGION` environment variable (and not `AWS_REGION` -- note that it's conspicuously missing from [here](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html)), while the golang SDK does _mostly_ the opposite -- it reads the region **only** from `AWS_REGION` and not at all `~/.aws/config`, **unless** the `AWS_SDK_LOAD_CONFIG` environment variable is set to a truthy value, in which case, it'll consider those (lower in precedence after reading `AWS_REGION`). Never mind the fact that the SDKs largely ignore the region choice when it comes to STS....

Anyway, importantly, in the golang SDK, you can also force the behavior of `AWS_SDK_LOAD_CONFIG` being set by calling `session.NewSessionWithOptions`. And, according to my testing, if it doesn't find something, the empty string will be returned as the region; there is no default. So, I think something like this would probably be best:

```go
session, err := session.NewSessionWithOptions(session.Options{SharedConfigState: session.SharedConfigEnable})
if err != nil {
        // should probably do something smarter here than a panic :)
        panic(err)
}
region := *session.Config.Region
if region == "" {
        // Go look in the metadata service
}
````

I'd suggest just using the same behavior here as in the agent. If people are impacted, it should probably fail the auto-unseal virtually immediately, and the fix is really simple -- either configure the right environment variables or delete the configured region in the KMS seal.

Thinking through this more thoroughly, I'm actually more worried about this potentially breaking the Vault agent auth. I suspect most people don't know about the weird regional-but-still-global nature of STS and it ignoring the region being set, plus the default in Vault being the global endpoint expecting a signature for us-east-1 probably means most people have Vault pointed at the default global endpoint. For people who use us-east-1 all the time anyway, this won't make a difference. But I can definitely see users who have config files or bash login profiles setting a different region by default (or who are running EC2 instances in another region, if you go looking in the instance metadata) now finding the Vault agent breaking on update because of this change. Ultimately I feel like this is still probably the right choice in the long run, but users will feel some pain given the way STS has evolved over the years and its current support. Maybe it'd be worth it to put together some additional documentation we can point users? If you think that would be helpful I can probably get something put together.

Reply:
I'm going to code it up in that manner, add some tests, and add some documentation, and will mark this as a non-draft PR when I've got all that pushed up. Thank you!